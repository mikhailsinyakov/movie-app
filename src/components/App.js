import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import { BrowserRouter as Router, Route } from "react-router-dom";
import { getMoviesList } from "api/movieAPI";
import { withLanguageContext } from "./Language";
import HomeScreen from "screens/Home";
import MovieScreen from "screens/Movie";

const moviesDataCache = {};

const App = ({ language }) => {
  const [loading, setLoading] = useState(false);
  const initialMoviesData = {
    page: 0,
    total_pages: 0,
    results: []
  };
  const [moviesData, setMoviesData] = useState(initialMoviesData);

  const fetchMoviesData = (page = 1) => {
    if (page === 1 && moviesDataCache[language]) {
      setMoviesData(moviesDataCache[language]);
    } else {
      setLoading(true);
      const { cancel, promise } = getMoviesList({ language, page });
      (async () => {
        try {
          const newMoviesData = await promise;
          const moviesDataToSave = {
            ...newMoviesData,
            results: [...moviesData.results, ...newMoviesData.results]
          };
          moviesDataCache[language] = moviesDataToSave;
          setMoviesData(moviesDataToSave);
          setLoading(false);
        } catch (e) {
          console.error(e);
        }
      })();
      return () => cancel("Component has unmounted");
    }
  };

  useEffect(fetchMoviesData, []);

  useEffect(() => {
    setMoviesData(initialMoviesData);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [language]);

  useEffect(() => {
    if (!moviesData.results.length) {
      return fetchMoviesData();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [moviesData]);

  return (
    <Router>
      <Route exact path="/">
        <HomeScreen
          moviesData={{
            list: moviesData.results,
            loading,
            completed: moviesData.page === moviesData.total_pages
          }}
          getMoreMovies={() => fetchMoviesData(moviesData.page + 1)}
        />
      </Route>
      <Route path="/movie/:id">
        <MovieScreen />
      </Route>
    </Router>
  );
};

App.propTypes = {
  language: PropTypes.string.isRequired
};

export default withLanguageContext(App);
