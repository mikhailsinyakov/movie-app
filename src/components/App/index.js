import React, { Component } from "react";

class App extends Component {
  render() {
    return <div>Movie App</div>;
  }
}

export default App;
