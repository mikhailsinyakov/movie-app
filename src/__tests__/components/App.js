import React from "react";
import Renderer from "react-test-renderer";
import { MemoryRouter as Router, Route } from "react-router-dom";
import {
  fireEvent,
  render,
  wait,
  waitForElementToBeRemoved
} from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import withLanguage from "../../components/Language";
import App from "components/App";

const Parent = ({ children }) => <div>{children}</div>;
const ParentWithLanguage = withLanguage(Parent);

describe("Snapshot", () => {
  let component;
  beforeEach(() => {
    component = Renderer.create(
      <Router>
        <ParentWithLanguage>
          <App />
        </ParentWithLanguage>
      </Router>
    );
  });
  afterEach(() => {
    component.unmount();
  });
  it("matches snapshot", () => {
    expect(component.toJSON()).toMatchSnapshot();
  });
});

const renderComponent = () =>
  render(
    <Router>
      <ParentWithLanguage>
        <App />
      </ParentWithLanguage>
    </Router>
  );

describe("Home page", () => {
  let component;
  beforeEach(() => (component = renderComponent()));
  afterEach(() => component.unmount());

  it("renders 'Loading...' text", () => {
    const { getByText } = component;
    expect(getByText(/loading/i)).toBeInTheDocument();
  });

  it("renders 20 images and 'More' button after a while", async () => {
    const { container, queryByText, getByText } = component;
    if (queryByText(/loading/i)) {
      await waitForElementToBeRemoved(() => queryByText(/loading/i));
    }
    expect(container.querySelectorAll("main img").length).toBe(20);
    expect(getByText(/more/i)).toBeInTheDocument();
  });

  it("renders 40 images after clicking 'More' button", async () => {
    const { container, queryByText, getByText } = component;
    if (queryByText(/loading/i)) {
      await waitForElementToBeRemoved(() => queryByText(/loading/i));
    }
    fireEvent.click(getByText(/more/i));
    await wait(() => expect(getByText(/loading/i)));
    await waitForElementToBeRemoved(() => getByText(/loading/i));
    expect(container.querySelectorAll("main img").length).toBe(40);
  });

  it("renders no 'More' button when has downloaded all movies", async () => {
    const { getByText, queryByText } = component;
    if (queryByText(/loading/i)) {
      await waitForElementToBeRemoved(() => queryByText(/loading/i));
    }
    let moreButton;
    moreButton = queryByText(/more/i);
    while (moreButton) {
      fireEvent.click(moreButton);
      await wait(() => expect(getByText(/loading/i)));
      await waitForElementToBeRemoved(() => getByText(/loading/i));
      moreButton = queryByText(/more/i);
    }
    expect(moreButton).toBeNull();
  });
});

describe("Movie page", () => {
  let component;
  beforeEach(() => (component = renderComponent()));
  afterEach(() => component.unmount());

  it("renders movie details page when clicking on poster", async () => {
    const { container, queryByText, getByAltText, queryByAltText } = component;
    if (queryByText(/loading/i)) {
      await waitForElementToBeRemoved(() => queryByText(/loading/i));
    }
    const firstPoster = container.querySelector("main img");
    const movieTitle = firstPoster.alt;
    fireEvent.click(firstPoster);
    expect(queryByAltText(movieTitle)).toBeNull();
    await wait(() => expect(getByAltText(movieTitle)));
    expect(getByAltText(movieTitle)).toBeInTheDocument();
  });
});
