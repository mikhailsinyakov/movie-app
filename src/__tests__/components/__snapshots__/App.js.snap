// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot matches snapshot 1`] = `
<div>
  <header
    className="sc-AxhCb hKLhMx"
  >
    <a
      className="sc-AxjAm eGEFpi"
      href="/"
      onClick={[Function]}
    >
      <img
        alt="app-logo"
        className="sc-AxirZ ggkxvc"
        src="img/app-logo.svg"
      />
    </a>
    <h1
      className="sc-AxiKw jstiRv"
    >
      Movie App
    </h1>
    <a
      className="sc-AxjAm eGEFpi"
      href="https://www.themoviedb.org/"
    >
      <img
        alt="tmdb-logo"
        className="sc-AxirZ loYNGT"
        src="img/tmdb-logo.png"
      />
    </a>
  </header>
  <main>
    <div
      className="sc-AxgMl ddxDUv"
    >
      <select
        className="sc-AxhUy lbQkcW"
        onChange={[Function]}
        value="en"
      >
        <option
          value="en"
        >
          EN
        </option>
        <option
          value="ru"
        >
          RU
        </option>
      </select>
    </div>
    <div
      className="sc-fzqNJr jJUNwy"
    >
      <div
        className="sc-fzoyAV gAOsEl"
      />
    </div>
  </main>
  <footer
    className="sc-fzoLag iMpqwM"
  >
    <span>
      This product uses the TMDb API but is not endorsed or certified by TMDb
    </span>
  </footer>
</div>
`;
